NSFS list_objects should only list latest objects

vh05 opened the pull request:
### Explain the changes
list_objects, returns every object under the bucket. It should ignore objects under .versions where we
store the versions of the object.

### Issues: Fixed #7164 

### Testing Instructions:
1. Create a object and create versions of it
2. do list_objects(), that should not return objects under `.versions`


- [ ] Doc added/updated
- [ ] Tests added

Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -1759,6 +1761,17 @@ class NamespaceFS {
+
+    _skip_version_dir(dir_key) {
+        const parts = dir_key.split('/');
+        let skip = false;
```
I think you can just return dir_key.indexOf(`/.${HIDDEN_VERSIONS_PATH}/`) > 0, no?

v-harihar replied:
Yeah sure. Need to be bit used to these methods :)

romayalon commented on src/test/unit_tests/test_namespace_fs.js:
```diff
@@ -232,6 +232,113 @@ mocha.describe('namespace_fs', function() {
+        });
+
+        mocha.after(async () => {
+            //fs_utils.folder_delete(tmp_fs_root);
```
please add it back for cleaning, can you please uncomment it in other tests in this file as well?

v-harihar replied:
that went unnoticed. Sure will do that

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -1759,6 +1761,13 @@ class NamespaceFS {
     }
+
+    _skip_version_dir(dir_key) {
+        if(dir_key.indexOf(`${HIDDEN_VERSIONS_PATH}`) >= 0) {
```
we expect at least 1 char before that so I think the condition should be > 0

v-harihar replied:
For `.versions` under bucket path will have 0 index. Ex: for  `$BUCKET_PATH/.versions` , the `dir_key` would be `.versions` that will have 0 index.

correct me if wrong

romayalon replied:
ah ok coool

guymguym commented on src/sdk/namespace_fs.js:
```diff
@@ -1759,6 +1761,10 @@ class NamespaceFS {
     }
+
+    _skip_version_dir(dir_key) {
+        return dir_key.indexOf(`${HIDDEN_VERSIONS_PATH}`) >= 0;
```
seems like the string interpolation (`${x}`) is unneeded 
```suggestion
        return dir_key.indexOf(HIDDEN_VERSIONS_PATH) >= 0;
```

romayalon replied:
shouldn't we search for /.versions/ ? instead of .versions

v-harihar replied:
Yes. In the previous code, the file parts were divided with `/` delimeter and here it should be `/.versions/`. I will correct it

romayalon replied:
Let's take for example the following dir:
dir_key = a/b/c/d/e/f/g/h/i/j/k/l/m/n/o/p/.versions
Definitions: 
n = dir_key.length
l = number of '/'
m = avg number of chars of every subdir
n = l + (l+1)*m

Option1 -
1. Splitting by '/' is 1 loop - O(n) - while n is the number of chars in dir_key.
2. Iterating the result array and comparing char by char of each item in the array O((l+1)*m) while m is the average number of chars of each item in the array and l is the number of '/' in the string = the number of items in the split result array.
Total complexity time - O(n+(l+1)*m)

Option 2 - 
1. searching char by char twice will take - O(2n)

So the question is (l+1)*m < n, the answer is yes and your assumption is correct, the difference is exactly l because we said that n = l + (l+1)*m
I think that it's a very slight improvement, I want to believe that l is not a big number in most of the cases unless I did a mistake in the calculation/ the algorithm of option 1, please correct me if you see a mistake.

Anyway, there is another optimization you can do for option 2 in order to iterate the string once - for example:
```
const idx = dir_key.indexOf(HIDDEN_VERSIONS_PATH);
return idx === 0 || idx > 0 && dir_key[idx-1] === ‘/’;
```
then the optimized version of option 2 is O(n).

v-harihar replied:
This is okay :)

aspandey replied:
@v-harihar do we have this version path like "dir_key = a/b/c/d/e/f/g/h/i/j/k/l/m/n/o/p/.versions" only?

If yes, why can't we first split this string, with "/" as delimiter, and then just compare last element with ".versions"?

I mean to say is that if we have any idea of the position of ".versions" then compare that only.

I know you would have thought of this but just mentioning.

romayalon replied:
So Vinayak is saying that dir_key can be exactly `.versions` (without '/') or `a/b/.versions` for example
I thought that we can use dir_key.endsWith('.versions') but still we need to check for the / if the index > 0 (that's why I wrote indexOf and not endsWith) / compare to .versions or search for dir_key.lastIndexOf('/') and then slice and compare to .versions

v-harihar replied:
@aspandey Yes logic wise both `split & compare` and the `indexOf() & check index (the added logic)` would work fine.  Also process cycle wise both results same. We had that logic earlier in this patch and overwritten because of append push which I should not have done :)

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -1759,6 +1761,11 @@ class NamespaceFS {
     }
+
+    _skip_version_dir(dir_key) {
+        const idx = dir_key.indexOf(HIDDEN_VERSIONS_PATH)
```
missing ;

vh05 committed:
list_objects, returns every object under the bucket.

It should ignore objects under .versions where we
store the versions of the object.

Issues: Fixed #7164

Testing Instructions:
1. Create a object and create versions of it
2. do list_objects(), that should not return objects under .versions

Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>
